src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1441,6 +1441,19 @@ uint MulAddS2INode::hash() const {
 
 //------------------------------Rotate Operations ------------------------------
 
+Node* RotateLeftNode::Identity(PhaseGVN* phase) {
+  const Type* t1 = phase->type(in(1));
+  if (t1 == Type::TOP) {
+    return this;
+  }
+  // Rotate by a multiple of 32/64 does nothing
+  int mask = (t1->isa_int() ? BitsPerJavaInteger : BitsPerJavaLong) - 1;
+  if ((getShiftCon(phase, this, -1) & mask) == 0) {
```

Comment:
I find the interaction between `in(2)`, `-1`, and `& mask == 0` intricate (it covers both cases: when shift constant is successfully extracted and when it is not).

Maybe refactor `getShiftCon()` to return a `bool` which signals that the constant was found and then return the value as an out argument?